File: src/safety/validator.ts
import { DEFAULT_INJECTION_PATTERNS, type InjectionPattern } from "./patterns.js";
import type { ValidatorLike } from "../types.js";
import type { ValidationIssue, ValidationResult } from "../types.js";

const MAX_INPUT_LENGTH = 100_000;

export class InputValidator implements ValidatorLike {
  public constructor(
    private readonly injectionPatterns: InjectionPattern[] = DEFAULT_INJECTION_PATTERNS,
  ) {}

  public getInjectionPatterns(): InjectionPattern[] {
    return this.injectionPatterns;
  }

  public validateInput(content: string, field: string = "input"): ValidationResult {
    const errors: ValidationIssue[] = [];
    const warnings: ValidationIssue[] = [];

    if (content.length === 0) {
      errors.push({ field, message: "Input cannot be empty", code: "empty_input", severity: "error" });
    }

    if (content.length > MAX_INPUT_LENGTH) {
      errors.push({ field, message: "Input exceeds maximum length", code: "input_too_long", severity: "error" });
    }

    for (const pattern of this.injectionPatterns) {
      if (pattern.regex.test(content)) {
        const issue: ValidationIssue = {
          field,
          message: `Potential prompt injection detected: ${pattern.name}`,
          code: `injection_${pattern.name}`,
          severity: pattern.action === "block" ? "error" : "warning",
        };
        if (pattern.action === "block") {
          errors.push(issue);
        } else {
          warnings.push(issue);
        }
      }
    }

    return { isValid: errors.length === 0, errors, warnings };
  }

  public validateToolParams(params: Record<string, unknown>): ValidationResult {
    const errors: ValidationIssue[] = [];
    const warnings: ValidationIssue[] = [];

    for (const [key, value] of Object.entries(params)) {
      if (typeof value === "string") {
        const result = this.validateInput(value, key);
        errors.push(...result.errors);
        warnings.push(...result.warnings);
      }
    }

    return { isValid: errors.length === 0, errors, warnings };
  }
}

File: titanx/safety/validator.py
from __future__ import annotations

from typing import Any

from .patterns import DEFAULT_INJECTION_PATTERNS, InjectionPattern
from ..types import ValidatorLike, ValidationIssue, ValidationResult

MAX_INPUT_LENGTH = 100_000


class InputValidator(ValidatorLike):
    def __init__(self, injection_patterns: list[InjectionPattern] | None = None) -> None:
        self._patterns = injection_patterns or DEFAULT_INJECTION_PATTERNS

    def get_injection_patterns(self) -> list[InjectionPattern]:
        return self._patterns

    def validate_input(self, content: str, field: str = "input") -> ValidationResult:
        errors: list[ValidationIssue] = []
        warnings: list[ValidationIssue] = []

        if not content:
            errors.append(ValidationIssue(field=field, message="Input cannot be empty", code="empty_input", severity="error"))

        if len(content) > MAX_INPUT_LENGTH:
            errors.append(ValidationIssue(field=field, message="Input exceeds maximum length", code="input_too_long", severity="error"))

        for pattern in self._patterns:
            if pattern.regex.search(content):
                issue = ValidationIssue(
                    field=field,
                    message=f"Potential prompt injection detected: {pattern.name}",
                    code=f"injection_{pattern.name}",
                    severity="error" if pattern.action == "block" else "warning",
                )
                (errors if pattern.action == "block" else warnings).append(issue)

        return ValidationResult(is_valid=len(errors) == 0, errors=errors, warnings=warnings)

    def validate_tool_params(self, params: dict[str, Any]) -> ValidationResult:
        errors: list[ValidationIssue] = []
        warnings: list[ValidationIssue] = []

        for key, value in params.items():
            if isinstance(value, str):
                result = self.validate_input(value, key)
                errors.extend(result.errors)
                warnings.extend(result.warnings)

        return ValidationResult(is_valid=len(errors) == 0, errors=errors, warnings=warnings)
